diff --git a/core/modules/migrate_drupal/src/Tests/Table/d7/Variable.php b/core/modules/migrate_drupal/src/Tests/Table/d7/Variable.php
index 8f4c3a8..b4a7396 100644
--- a/core/modules/migrate_drupal/src/Tests/Table/d7/Variable.php
+++ b/core/modules/migrate_drupal/src/Tests/Table/d7/Variable.php
@@ -338,6 +338,9 @@ public function load() {
       'name' => 'theme_default',
       'value' => 's:6:"bartik";',
     ))->values(array(
+      'name' => 'tracker_batch_size',
+      'value' => 'i:999;',
+    ))->values(array(
       'name' => 'user_admin_role',
       'value' => 's:1:"3";',
     ))->values(array(
@@ -449,4 +452,4 @@ public function load() {
   }
 
 }
-#5b6552f715939e2b33c22779e47e73e3
+#a29150ffe905f457be20f5a287fdcef0
diff --git a/core/modules/tracker/migration_templates/d7_tracker_settings.yml b/core/modules/tracker/migration_templates/d7_tracker_settings.yml
new file mode 100644
index 0000000..269c2ea
--- /dev/null
+++ b/core/modules/tracker/migration_templates/d7_tracker_settings.yml
@@ -0,0 +1,13 @@
+id: d7_tracker_settings
+label: Drupal 7 tracker settings
+migration_tags:
+  - Drupal 7
+source:
+  plugin: variable
+  variables:
+    - tracker_batch_size
+process:
+  cron_index_limit: tracker_batch_size
+destination:
+  plugin: config
+  config_name: tracker.settings
diff --git a/core/modules/tracker/src/Tests/Migrate/d7/MigrateTrackerSettingsTest.php b/core/modules/tracker/src/Tests/Migrate/d7/MigrateTrackerSettingsTest.php
new file mode 100644
index 0000000..3dd67b3
--- /dev/null
+++ b/core/modules/tracker/src/Tests/Migrate/d7/MigrateTrackerSettingsTest.php
@@ -0,0 +1,38 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\tracker\Tests\Migrate\d7\MigrateTrackerSettingsTest.
+ */
+
+namespace Drupal\tracker\Tests\Migrate\d7;
+
+use Drupal\migrate_drupal\Tests\d7\MigrateDrupal7TestBase;
+
+/**
+ * Tests migration of Tracker settings to configuration.
+ *
+ * @group tracker
+ */
+class MigrateTrackerSettingsTest extends MigrateDrupal7TestBase {
+
+  public static $modules = ['tracker'];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+    $this->installConfig(['tracker']);
+    $this->loadDumps(['Variable.php']);
+    $this->executeMigration('d7_tracker_settings');
+  }
+
+  /**
+   * Tests migration of tracker's variables to configuration.
+   */
+  public function testMigration() {
+    $this->assertIdentical(999, \Drupal::config('tracker.settings')->get('cron_index_limit'));
+  }
+
+}
