diff --git a/core/modules/views/src/Plugin/views/style/Mapping.php b/core/modules/views/src/Plugin/views/style/Mapping.php
index e4823e6..b647ea9 100644
--- a/core/modules/views/src/Plugin/views/style/Mapping.php
+++ b/core/modules/views/src/Plugin/views/style/Mapping.php
@@ -103,7 +103,8 @@ public function buildOptionsForm(&$form, FormStateInterface $form_state) {
       // Optionally filter the available fields.
       if (isset($mapping[$key]['#filter'])) {
         $this->view->initHandlers();
-        $this::$mapping[$key]['#filter']($field_options);
+        $filter = $mapping[$key]['#filter'];
+        $this::$filter($field_options);
         unset($mapping[$key]['#filter']);
       }
 
