diff --git a/core/lib/Drupal/Core/Language/LanguageManager.php b/core/lib/Drupal/Core/Language/LanguageManager.php
index cd27c74..2fd1597 100644
--- a/core/lib/Drupal/Core/Language/LanguageManager.php
+++ b/core/lib/Drupal/Core/Language/LanguageManager.php
@@ -207,6 +207,7 @@ public function getDefaultLockedLanguages($weight = 0) {
     $locked_language = array(
       'default' => FALSE,
       'locked' => TRUE,
+      'direction' => LanguageInterface::DIRECTION_LTR,
     );
     // This is called very early while initializing the language system. Prevent
     // early t() calls by using the TranslationWrapper.
