diff --git a/core/lib/Drupal/Core/Database/Query/SelectInterface.php b/core/lib/Drupal/Core/Database/Query/SelectInterface.php
index 6b87014..cd49505 100644
--- a/core/lib/Drupal/Core/Database/Query/SelectInterface.php
+++ b/core/lib/Drupal/Core/Database/Query/SelectInterface.php
@@ -147,6 +147,9 @@ public function escapeLike($string);
    * For some database drivers, it may also wrap the field name in
    * database-specific escape characters.
    *
+   * @param string $string
+   *   An unsanitized field name.
+   *
    * @return
    *   The sanitized field name string.
    */
