diff --git a/core/modules/migrate/src/Plugin/migrate/destination/EntityRevision.php b/core/modules/migrate/src/Plugin/migrate/destination/EntityRevision.php
index 8aaefef..ea4c9d5 100644
--- a/core/modules/migrate/src/Plugin/migrate/destination/EntityRevision.php
+++ b/core/modules/migrate/src/Plugin/migrate/destination/EntityRevision.php
@@ -9,6 +9,7 @@
 
 use Drupal\Core\Entity\ContentEntityInterface;
 use Drupal\migrate\MigrateException;
+use Drupal\migrate\MigrateSkipRowException;
 use Drupal\migrate\Row;
 
 /**
@@ -44,6 +45,12 @@ protected function getEntity(Row $row, array $old_destination_id_values) {
     else {
       $entity_id = $row->getDestinationProperty($this->getKey('id'));
       $entity = $this->storage->load($entity_id);
+
+      // If we fail to load the original entity revision something is wrong.
+      if (!$entity) {
+        throw new MigrateSkipRowException();
+      }
+
       $entity->enforceIsNew(FALSE);
       $entity->setNewRevision(TRUE);
     }
